### PR TITLE
[Enhancement] Support tvm

### PR DESCRIPTION
**Pull request by grimoire**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add TVM support.

## Modification

- TVM builder
- AUTOTVM and ANSOR Tuner support
- SDK support

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by grimoire** [comment]
Please move all cuda kernel to `common/cuda/` as .cuh and include it in .cu.

File: csrc/mmdeploy/device/cpu/cpu_device.cpp
Comment (tpoisonooo) [inline csrc/mmdeploy/device/cpu/cpu_device.cpp:58]: Unaligned data almost useless for deep learning inference, how about remove `data_` and keep it aligned by default.

File: csrc/mmdeploy/device/cpu/cpu_device.cpp
Comment (tpoisonooo) [inline csrc/mmdeploy/device/cpu/cpu_device.cpp:14]: `space = (size + alignment - 1)  &  ~(alignment - 1)`

If mmdeploy depends on c++17, use `std::aligned_alloc`  https://en.cppreference.com/w/cpp/memory/c/aligned_alloc

File: csrc/mmdeploy/device/cuda/cuda_device.cpp
Comment (tpoisonooo) [inline csrc/mmdeploy/device/cuda/cuda_device.cpp:72]: How about giving a  default value,  caller does not know moderate `alignment` value.

File: docs/en/03-benchmark/benchmark_tvm.md
Comment (tpoisonooo) [inline docs/en/03-benchmark/benchmark_tvm.md:39]: ViT precision drop is bigger than`1%`, I think user need explanation.

**Comment by tpoisonooo** [comment]
1. missing `how_to_build.md`,  I do not know how to install tvm after reading mmdeploy doc.
2. how about adding script install tvm backend
3. missing tvm CI

**Comment by tpoisonooo** [comment]
LGTM after fix review advices.

**Review by tpoisonooo — APPROVED** [review]

File: csrc/mmdeploy/device/cpu/cpu_device.cpp
Comment (grimoire) [reply csrc/mmdeploy/device/cpu/cpu_device.cpp:58]: CpuHostMemory accepts to build from raw ptr.
https://github.com/open-mmlab/mmdeploy/blob/e28e47128d94315dc1099562925d9da992d47a72/csrc/mmdeploy/device/cpu/cpu_device.cpp#L35-L40

The data is not owned by the class and we can not guarantee that the data is aligned.

File: csrc/mmdeploy/core/tensor.h
Comment (lzhangzz) [inline csrc/mmdeploy/core/tensor.h:70]: Move dlpack stuff to csrc/mmdeploy/utils and refactor them as free functions.

File: csrc/mmdeploy/device/cpu/cpu_device.cpp
Comment (grimoire) [reply csrc/mmdeploy/device/cpu/cpu_device.cpp:14]: https://github.com/open-mmlab/mmdeploy/actions/runs/3407199937/jobs/5666603714#step:5:69
https://en.cppreference.com/w/cpp/algorithm/exclusive_scan
SDK build failed with exclusive_scan, I guess SDK is build with c++14.

File: csrc/mmdeploy/core/tensor.h
Comment (lzhangzz) [inline csrc/mmdeploy/core/tensor.h:18]: Remove strides support and simply set `DLTensor::strides` to `nullptr`. Adding strides before we have a detailed plan to support strides is dangerous. As the rest of the codebase does not support strided tensors, `is_contiguous` must be checked everywhere. It also enlarges `Tensor` size and complicates tensor opreations.

File: csrc/mmdeploy/net/tvm/tvm_net.cpp
Comment (lzhangzz) [inline csrc/mmdeploy/net/tvm/tvm_net.cpp:218]: The TVM vm seems to be running on the default stream. While `stream_` is not the CUDA default stream. The `Wait` has no effect here.

May use 
```C++
if (device_.is_device()) {
  OUTCOME_TRY(Stream(device_, nullptr).Wait());
}
```
to construct a wrapper on default stream and wait for it.

**Review by lzhangzz — APPROVED** [review]
LGTM

**Comment by grimoire** [comment]
Do not merge now. find bugs in sdk.

**Comment by lvhan028** [comment]
get an error in `ci/circleci: test_windows_onnxruntime`, but I haven't figured it out.

**Comment by lvhan028** [comment]
@irexyc

File: csrc/mmdeploy/device/cpu/cpu_device.cpp
Comment (irexyc) [inline csrc/mmdeploy/device/cpu/cpu_device.cpp:15]: msvc doesn't have std::aligned_alloc https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/aligned-malloc?view=msvc-170

may use _aligned_malloc/_aligned_free instead on windows

File: cmake/modules/FindTVM.cmake
Comment (irexyc) [inline cmake/modules/FindTVM.cmake:28]: lack PATH_SUFFIXES for windows. If we build tvm on ${TVM_DIR}/build, we may add `build/Release` or `build/Debug`.

Otherwise, we should make it clear in the document that the tvm should be installed in ${TVM_DIR}/build dir

File: cmake/modules/FindTVM.cmake
Comment (lvhan028) [inline cmake/modules/FindTVM.cmake:28]: May set `build/Release` and `build/Debug` according to `CMAKE_BUILD_TYPE`

**Review by lvhan028 — APPROVED** [review]

File: csrc/mmdeploy/net/tvm/tvm_net.cpp
Comment (lvhan028) [reply csrc/mmdeploy/net/tvm/tvm_net.cpp]: what's the status now?

**Pull request merged**
